src/storage/nats_kv.rs

```diff
@@ -0,0 +1,304 @@
+};
+
+use tokio::fs::read as read_file;
+
+use derive_builder::Builder;
+use semver::{BuildMetadata, Prerelease, Version};
+use serde::ser::StdError;
+use serde::{Deserialize, Serialize};
+use serde_json::{from_slice, to_vec, Error as SerdeJsonError};
+
+use super::{EngineMetadata, LatticeStorage, StorageEngineError, StorageError, StoreOptions};
+
+use super::state::{LatticeState, LatticeStateBuilder, LatticeStateBuilderError};
+
+use crate::storage::{Store, StoreRead, StoreReadWrite, StoreWrite, WithStateMetadata};
+
+const STORAGE_ENGINE_NAME: &str = "nats_kv";
+const STORAGE_ENGINE_VERSION: Version = Version {
+    major: 0,
+    minor: 1,
+    patch: 0,
+    pre: Prerelease::EMPTY,
+    build: BuildMetadata::EMPTY,
+};
+
+////////////
+// Engine //
+////////////
+
+#[derive(Debug)]
+pub struct NatsKvStorageEngine {
+    /// Configuration for the NATS KV storage engine
+    config: NatsKvStorageConfig,
+
+    /// NATS client
+    client: NatsClient,
+}
+
+#[derive(Debug, Serialize, Deserialize, Builder)]
+pub struct NatsKvStorageConfig {
+    /// URL to the nats instance
+    pub nats_url: String,
+
+    /// Prefix to use with lattice buckets
+    pub lattice_bucket_prefix: Option<String>,
+
+    /// Authentication config for use with NATS
+    pub auth: Option<NatsAuthConfig>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct NatsAuthConfig {
+    pub creds_file: Option<String>,
+    pub jwt_seed: Option<String>,
+    pub jwt_path: Option<PathBuf>,
+}
+
+impl From<nkeys::error::Error> for StorageEngineError {
+    fn from(err: nkeys::error::Error) -> StorageEngineError {
+        StorageEngineError::Engine(format!("failed to process nkeys: {err}"))
+    }
+}
+
+impl From<std::io::Error> for StorageEngineError {
+    fn from(err: std::io::Error) -> StorageEngineError {
+        StorageEngineError::Engine(format!("unexpected I/O error: {err}"))
+    }
+}
+
+async fn build_nats_options(
+    cfg: &NatsAuthConfig,
+) -> Result<NatsConnectOptions, StorageEngineError> {
+    // Use JWT + seed if present
+    if let (Some(path), Some(seed)) = (&cfg.jwt_path, &cfg.jwt_seed) {
+        let file_contents = read_file(path).await?;
+        let jwt = String::from_utf8_lossy(&file_contents);
+        let kp = std::sync::Arc::new(nkeys::KeyPair::from_seed(seed)?);
+        return Ok(async_nats::ConnectOptions::with_jwt(
+            jwt.into(),
+            move |nonce| {
+                let key_pair = kp.clone();
+                async move { key_pair.sign(&nonce).map_err(async_nats::AuthError::new) }
+            },
+        ));
+    }
+
+    // Use creds file if present
+    if let Some(path) = &cfg.creds_file {
+        return NatsConnectOptions::with_credentials_file(path.clone().into())
+            .await
+            .map_err(|e| StorageEngineError::Engine(format!("Failed to connect NATS: {e}")));
+    }
+
+    Ok(NatsConnectOptions::default())
+}
+
+impl NatsKvStorageEngine {
+    #[allow(dead_code)]
```

Comment:
I assume this was for testing? Should probably be removed now that this is `pub`